ansible/plugins/connection/multi_passwd_ssh.py

```diff
@@ -37,6 +47,38 @@ def _password_retry(func):
     """
     @wraps(func)
     def wrapped(self, *args, **kwargs):
+
+        # If the host have an IPv6 address, try connect IPv4 address first,
+        # If IPv4 host unavailable, fall back to use IPv6 address
+        try:
+            hostv6 = self.get_option("hostv6")
+        except KeyError:
+            hostv6 = None
+
+        if hostv6:
+            try:
+                return func(self, *args, **kwargs)
+            except AnsibleConnectionFailure as e:
+                logger.info("First connection failed: {}".format(str(e)))
+                if CONNECTION_TIMEOUT_ERR_FLAG in e.message:
+                    self._play_context.remote_addr = hostv6
+                    # args sample:
+                    # ( [b'sshpass', b'-d18', b'ssh', b'-o', b'ControlMaster=auto', b'-o', b'ControlPersist=120s', b'-o', b'UserKnownHostsFile=/dev/null', b'-o', b'StrictHostKeyChecking=no', b'-o', b'StrictHostKeyChecking=no', b'-o', b'User="admin"', b'-o', b'ConnectTimeout=60', b'-o', b'ControlPath="/home/jianquanye/.ansible/cp/376bdcc730"', '2603:10e2:140:3000::96', b'/bin/sh -c \'echo PLATFORM; uname; echo FOUND; command -v \'"\'"\'python3.10\'"\'"\'; command -v \'"\'"\'python3.9\'"\'"\'; command -v \'"\'"\'python3.8\'"\'"\'; command -v \'"\'"\'python3.7\'"\'"\'; command -v \'"\'"\'python3.6\'"\'"\'; command -v \'"\'"\'python3.5\'"\'"\'; command -v \'"\'"\'/usr/bin/python3\'"\'"\'; command -v \'"\'"\'/usr/libexec/platform-python\'"\'"\'; command -v \'"\'"\'python2.7\'"\'"\'; command -v \'"\'"\'/usr/bin/python\'"\'"\'; command -v \'"\'"\'python\'"\'"\'; echo ENDFOUND && sleep 0\''], None) # noqa: E501
+                    # args[0] are the parameters of ssh connection
+                    ssh_args = args[0]
+                    # Change the IPv4 host in the ssh_args to IPv6
+                    for idx in range(len(ssh_args)):
+                        if type(ssh_args[idx]) == bytes and ssh_args[idx].decode() == self.host:
+                            ssh_args[idx] = hostv6
+                    self.host = hostv6
+                    self.set_option("host", hostv6)
+            except BaseException as e:
+                # Only catch the connection error, won't block the multi-password functionality
+                logger.info("First connection failed: {}".format(str(e)))
+
+            # Reset the sshpass_pipe for the new connections to be created
+            self.sshpass_pipe = os.pipe()
+
```

Comment:
Can this enhancement work together with IPv6 only and altpasswords?

Reply:
Yes, they can co-work properly and independently,
I tried to manually modify the pwd of DUT to altpasswords, and and the IPv6 convert and revert works well.